automod/engine/engine.go

```diff
@@ -0,0 +1,169 @@
+		}
+	}()
+
+	ident, err := eng.Directory.LookupDID(ctx, did)
+	if err != nil {
+		return fmt.Errorf("resolving identity: %w", err)
+	}
+	if ident == nil {
+		return fmt.Errorf("identity not found for did: %s", did.String())
+	}
+
+	am, err := eng.GetAccountMeta(ctx, ident)
+	if err != nil {
+		return err
+	}
+	ac := NewAccountContext(ctx, eng, *am)
+	if err := eng.Rules.CallIdentityRules(&ac); err != nil {
+		return err
+	}
+	eng.CanonicalLogLineAccount(&ac)
+	eng.PurgeAccountCaches(ctx, am.Identity.DID)
+	if err := eng.persistAccountModActions(&ac); err != nil {
+		return err
+	}
+	if err := eng.persistCounters(ctx, &ac.effects); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (eng *Engine) ProcessRecordOp(ctx context.Context, op RecordOp) error {
+	// similar to an HTTP server, we want to recover any panics from rule execution
+	defer func() {
+		if r := recover(); r != nil {
+			eng.Logger.Error("automod event execution exception", "err", r, "did", op.DID, "collection", op.Collection, "rkey", op.RecordKey)
+		}
+	}()
+
+	if err := op.Validate(); err != nil {
+		return fmt.Errorf("bad record op: %w", err)
+	}
+	ident, err := eng.Directory.LookupDID(ctx, op.DID)
+	if err != nil {
+		return fmt.Errorf("resolving identity: %w", err)
+	}
+	if ident == nil {
+		return fmt.Errorf("identity not found for did: %s", op.DID)
+	}
+
+	am, err := eng.GetAccountMeta(ctx, ident)
+	if err != nil {
+		return err
+	}
+	rc := NewRecordContext(ctx, eng, *am, op)
+	rc.Logger.Debug("processing record")
+	switch op.Action {
+	case CreateOp, UpdateOp:
+		if err := eng.Rules.CallRecordRules(&rc); err != nil {
+			return err
+		}
+	case DeleteOp:
+		if err := eng.Rules.CallRecordDeleteRules(&rc); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("unexpected op action: %s", op.Action)
+	}
+	eng.CanonicalLogLineRecord(&rc)
+	// purge the account meta cache when profile is updated
+	if rc.RecordOp.Collection == "app.bsky.actor.profile" {
+		eng.PurgeAccountCaches(ctx, am.Identity.DID)
+	}
+	if err := eng.persistRecordModActions(&rc); err != nil {
+		return err
+	}
+	if err := eng.persistCounters(ctx, &rc.effects); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *Engine) GetCount(name, val, period string) (int, error) {
+	return e.Counters.GetCount(context.TODO(), name, val, period)
+}
+
+func (e *Engine) GetCountDistinct(name, bucket, period string) (int, error) {
+	return e.Counters.GetCountDistinct(context.TODO(), name, bucket, period)
+}
+
+// checks if `val` is an element of set `name`
+func (e *Engine) InSet(name, val string) (bool, error) {
+	return e.Sets.InSet(context.TODO(), name, val)
+}
+
+// purge caches of any exiting metadata
+func (e *Engine) PurgeAccountCaches(ctx context.Context, did syntax.DID) error {
+	e.Directory.Purge(ctx, did.AtIdentifier())
+	return e.Cache.Purge(ctx, "acct", did.String())
+}
+
+func (e *Engine) CanonicalLogLineAccount(c *AccountContext) {
```

Comment:
Agree that `CanonicalLogLineAccount` makes sense to have as an engine method since there's no reason to let other code control (and thus be able to forget to) call it.  The other argument would be that it mostly sees fields of effects, but I think log methods being closer to their use than their subject makes sense.

I think it can even be private?  There appears to be one caller, and it is also on the engine.

Reply:
:+1: to having it be private, didn't think of that.

A possible refactor will be to have an "alerts" or "notifications" component/interface on engine, which gets called after rule processing to handle things like sending Slack messages or logging more loudly for some cases. I *think* that we'll always want basic logging ("canonical log line") to happen regardless though, and can be done by the engine.